src/main/java/dev/koifysh/archipelago/Client.java

```diff
@@ -117,45 +141,87 @@ public void removeTag(String tag) {
+            HashMap<String,File> localGamesList = new HashMap<String,File>();
 
-        } catch (IOException e) {
-            LOGGER.info("no dataPackage found creating a new one.");
-            dataPackage = new DataPackage();
-            saveDataPackage();
-        } catch (ClassNotFoundException e) {
-            LOGGER.info("Failed to Read Previous datapackage. Creating new one.");
-            dataPackage = new DataPackage();
+            for(File gameDir : directoryPath.listFiles()){
+                localGamesList.put(gameDir.getName(), gameDir);
+            }
+
+            if(!localGamesList.isEmpty()){
+                for(String gameName : games){
+                    if(localGamesList.containsKey(gameName) && localGamesList.get(gameName).isDirectory()){
+                        //check all checksums
+                        for(File version : localGamesList.get(gameName).listFiles()){
+                            if(versions.containsKey(gameName) && versions.get(gameName).equals(version.getName())){
+                                try(FileReader reader = new FileReader(version)){
+                                    updateDataPackage(gson.fromJson(reader, DataPackage.class));
+                                    LOGGER.info("Read datapackage for Game: ".concat(gameName).concat(" Checksum: ").concat(version.getName()));
+                                } catch (IOException e){
+                                    LOGGER.info("Failed to read a datapackage. Creating a new one.");
+                                    dataPackage = new DataPackage();
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        } else{
+            //cache doesn't exist. Create the filepath
+            boolean success = directoryPath.mkdirs();
+            if(success){
+                //make a new datapackage
+                LOGGER.info("DataPackage directory didn't exist. Starting from fresh.");
+                dataPackage = new DataPackage();
+                //saveDataPackage();
+            } else{
+                //failed to make directories?
+                LOGGER.severe("Failed to make directories for datapackage cache.");
+            }
         }
     }
 
-    void saveDataPackage() {
-        try {
-            File dataPackageFile = new File(dataPackageLocation);
-
-            //noinspection ResultOfMethodCallIgnored
-            dataPackageFile.getParentFile().mkdirs();
-            //noinspection ResultOfMethodCallIgnored
-            dataPackageFile.createNewFile();
-
-            FileOutputStream fileOut = new FileOutputStream(dataPackageFile);
-            ObjectOutputStream objectOut = new ObjectOutputStream(fileOut);
-
-            objectOut.writeObject(dataPackage);
-
-
-            fileOut.close();
-            objectOut.close();
+    public void saveDataPackage() {
```

Comment:
```suggestion
    public synchronized void saveDataPackage() {
```

Reply:
This will lock inside a single process.  This will not account for cross process collisions.